ddtrace/pin.py

```diff
@@ -104,25 +117,33 @@ def onto(self, obj, send=True):
                 return obj.__setddpin__(self)
 
             pin_name = _DD_PIN_PROXY_NAME if isinstance(obj, wrapt.ObjectProxy) else _DD_PIN_NAME
+
+            # set the target reference; any get_from, clones and retarget the new PIN
+            self._target = obj
             return setattr(obj, pin_name, self)
         except AttributeError:
             log.debug("can't pin onto object. skipping", exc_info=True)
 
     def clone(self, service=None, app=None, app_type=None, tags=None, tracer=None):
-        """ Return a clone of the pin with the given attributes replaced. """
+        """Return a clone of the pin with the given attributes replaced."""
+        # do a shallow copy of Pin dicts
         if not tags and self.tags:
-            # do a shallow copy of the tags if needed.
-            tags = {k:v for k, v in self.tags.items()}
+            tags = self.tags.copy()
+
+        config = self._config.copy()
```

Comment:
very fast operation, considering our settings have a limited cardinality (<10)

Reply:
Does it need to be a `deepcopy`? Don't we risk to have one day some nested config?

Reply:
This is something we have to consider, but here some benchmarks: 
```python
# from Django config, just to simulate a small set of configurations with an empty dict
DEFAULTS = {'AGENT_HOSTNAME': 'localhost',
 'AGENT_PORT': 8126,
 'AUTO_INSTRUMENT': True,
 'DEFAULT_DATABASE_PREFIX': '',
 'DEFAULT_SERVICE': 'django',
 'DISTRIBUTED_TRACING': False,
 'ENABLED': True,
 'INSTRUMENT_CACHE': True,
 'INSTRUMENT_DATABASE': True,
 'INSTRUMENT_TEMPLATE': True,
 'TAGS': {},
 'TRACER': 'ddtrace.tracer'
}
# test 1
DEFAULTS.copy()
# test 2
copy.deepcopy(DEFAULTS)
```
And here the results:
```
test 1: 0.00654911994934082
test 2: 0.2787208557128906
```
The huge overhead is because of the string copy, so if nested dictionaries are used and we go with a `deepcopy`, we should take in consideration such overhead.

What do you think?

Reply:
Yeah that's a big difference. If for now `copy` is enough for all configuration options we want to provide, no reason to do more.

Reply:
OK I will add a comment to explain why we use a `copy` for our use case